packages/native/types/base.d.ts

```diff
@@ -0,0 +1,227 @@
+// Definitions by: Pat Sissons <https://github.com/patsissons>
+// TypeScript Version: 3.4
+
+import {
+  ComponentPropsWithoutRef,
+  ComponentType,
+  NamedExoticComponent
+} from 'react'
+import * as RN from 'react-native'
+
+type ReactNative = typeof RN
+
+export type ReactNativeStyle = RN.ViewStyle | RN.TextStyle | RN.ImageStyle
+
+export type ReactNativeComponentNames =
+  | 'ActivityIndicator'
+  | 'ActivityIndicatorIOS'
+  | 'Button'
+  | 'DatePickerIOS'
+  | 'DrawerLayoutAndroid'
+  | 'FlatList'
+  | 'Image'
+  | 'ImageBackground'
+  | 'KeyboardAvoidingView'
+  | 'ListView'
+  | 'Modal'
+  | 'NavigatorIOS'
+  | 'Picker'
+  | 'PickerIOS'
+  | 'ProgressBarAndroid'
+  | 'ProgressViewIOS'
+  | 'RecyclerViewBackedScrollView'
+  | 'RefreshControl'
+  | 'SafeAreaView'
+  | 'ScrollView'
+  | 'SectionList'
+  | 'SegmentedControlIOS'
+  | 'Slider'
+  | 'SnapshotViewIOS'
+  | 'StatusBar'
+  | 'SwipeableListView'
+  | 'Switch'
+  | 'SwitchIOS'
+  | 'TabBarIOS'
+  | 'Text'
+  | 'TextInput'
+  | 'ToolbarAndroid'
+  | 'TouchableHighlight'
+  | 'TouchableNativeFeedback'
+  | 'TouchableOpacity'
+  | 'TouchableWithoutFeedback'
+  | 'View'
+  | 'ViewPagerAndroid'
+  | 'WebView'
+
+export type ReactNativeComponents = Pick<ReactNative, ReactNativeComponentNames>
+
+export type ReactNativeComponentProps<
+  ComponentName extends ReactNativeComponentNames
+> = ComponentPropsWithoutRef<ReactNativeComponents[ComponentName]>
+
+export type ReactNativeStyleType<Props> = Props extends {
+  style?: RN.StyleProp<infer StyleType>
+}
+  ? StyleType
+  : ReactNativeStyle
```

Comment:
this construct lets us extract the specific shape of a `react-native` style prop, and support custom components too.